lib/src/matrix/decomposition/singular_value.dart

```diff
@@ -1,483 +1,847 @@
-        _v = Matrix(DataType.float, input.colCount, input.colCount),
-        _s = DataType.float
-            .newList(math.min(input.rowCount + 1, input.colCount)),
-        _m = input.rowCount,
-        _n = input.colCount {
-    // Initialize.
-    final A = input.cast(DataType.float).toMatrix();
-    final e = DataType.float.newList(_n);
-    final work = DataType.float.newList(_m);
-
-    // Reduce A to bi-diagonal form, storing the diagonal elements
-    // in s and the super-diagonal elements in e.
-    final nct = math.min(_m - 1, _n);
-    final nrt = math.max(0, math.min(_n - 2, _m));
-    for (var k = 0; k < math.max(nct, nrt); k++) {
-      if (k < nct) {
-        // Compute the transformation for the k-th column and
-        // place the k-th diagonal in s[k].
-        // Compute 2-norm of k-th column without under/overflow.
-        _s[k] = 0.0;
-        for (var i = k; i < _m; i++) {
-          _s[k] = hypot(_s[k], A.getUnchecked(i, k));
+  /// Initializes a new instance of the [SingularValueDecomposition].
+  SingularValueDecomposition._(
+      this._s, this._u, this._vt, this.vectorsComputed) {
+    _lazyW = computeW();
+  }
+
+  /// Initializes a new instance of the [SingularValueDecomposition].
+  /// This object will compute the singular value decomposition
+  /// when the constructor is called and cache it's decomposition.
+  factory SingularValueDecomposition(Matrix<num> matrix,
+      [bool computeVectors = true]) {
+    final nm = min(matrix.rowCount, matrix.colCount);
+
+    final uValues = DataType.float64.newList(matrix.rowCount * matrix.rowCount);
+    final sValues = DataType.float64.newList(nm);
+    final vtValues =
+        DataType.float64.newList(matrix.colCount * matrix.colCount);
+
+    _singularValueDecomposition(computeVectors, _columnMajorValuesOf(matrix),
+        matrix.rowCount, matrix.colCount, sValues, uValues, vtValues);
+
+    final u = Matrix<double>.fromPackedColumns(
+        DataType.float64, matrix.rowCount, matrix.rowCount, uValues,
+        format: MatrixFormat.columnMajor);
+    final s = Vector<double>.fromList(DataType.float64, sValues);
+    final vt = Matrix<double>.fromPackedColumns(
+        DataType.float64, matrix.colCount, matrix.colCount, vtValues,
+        format: MatrixFormat.columnMajor);
+
+    return SingularValueDecomposition._(s, u, vt, computeVectors);
+  }
+
+  late Vector<double> _s;
+  late Matrix<double> _u;
+  late Matrix<double> _vt;
+  late Matrix<double> _lazyW;
+
+  Matrix<double> computeW() {
+    final rows = _u.rowCount;
+    final columns = _vt.colCount;
+    final result = Matrix<double>(DataType.float64, rows, columns);
+    for (var i = 0; i < rows; i++) {
+      for (var j = 0; j < columns; j++) {
+        if (i == j) {
+          result.setUnchecked(i, i, _s[i]);
         }
-        if (_s[k] != 0.0) {
-          if (A.getUnchecked(k, k) < 0.0) {
-            _s[k] = -_s[k];
+      }
+    }
+    return result;
```

Comment:
I think this could be `_s.diagonalMatrix`?

Reply:
For a given m x n matrix, `s` is `min(m, n)` vector, however, `w` is `m x n` rectangular matrix, in which diagonal elements are singular values `s`.

A given `m x n` matrix `M` can be decomposed by `M = U W V*` where `U = m x m`, `W = m x n`, and `V = n x n` matrix. (V* or VT is the transpose of V.)

Reply:
Updated